cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -681,12 +788,77 @@ class row_hasher;
 }
 
 namespace equality {
+
+/**
+ * @brief Equality comparator functor that compares physical values rather than logical
+ * elements like lists, strings, or structs. It evaluates `NaN` not equal to all other values for
+ * IEEE-754. compliance.
+ */
+struct physical_equality_comparator {
+  /**
+   * @brief Operator for equality comparisons.
+   *
+   * Note that `NaN != NaN`, following IEEE-754.
+   *
+   * @param lhs First element
+   * @param rhs Second element
+   * @return `true` if `lhs == rhs` else `false`
+   */
+  template <typename Element>
+  __device__ constexpr bool operator()(Element const lhs, Element const rhs) const noexcept
+  {
+    return lhs == rhs;
+  }
+};
+
+/**
+ * @brief Equality comparator functor that compares physical values rather than logical
+ * elements like lists, strings, or structs. It evaluates `NaN` as equal to other `NaN`s.
+ */
+struct nan_equal_physical_equality_comparator {
+  /**
+   * @brief Operator for equality comparison of non-floating point values.
+   *
+   * @param lhs First element
+   * @param rhs Second element
+   * @return `true` if `lhs == rhs` else `false`
+   */
+  template <typename Element, CUDF_ENABLE_IF(not std::is_floating_point_v<Element>)>
+  __device__ constexpr bool operator()(Element const lhs, Element const rhs) const noexcept
+  {
+    return lhs == rhs;
+  }
+
+  /**
+   * @brief Operator for equality comparison of floating point values.
+   *
+   * Note that `NaN == NaN`.
+   *
+   * @param lhs First element
+   * @param rhs Second element
+   * @return `true` if `lhs` == `rhs` else `false`
+   */
+  template <typename Element, CUDF_ENABLE_IF(std::is_floating_point_v<Element>)>
+  __device__ constexpr bool operator()(Element const lhs, Element const rhs) const noexcept
+  {
+    return isnan(lhs) and isnan(rhs) ? true : lhs == rhs;
+  }
+};
+
 /**
- * @brief Comparator for performing equality comparison between the rows of two tables.
+ * @brief Computes the equality comparison between 2 rows.
+ *
+ * Equality is determined by:
+ * - Two rows are compared element by element.
+ * - The first mismatching element returns false representing unequal rows
+ * - If the rows are compared without mismatched elements, the rows are equivalent
```

Comment:
Prose feels more natural here, since the bullets aren't really "steps" being followed. Note that I used "equal" instead of "equivalent." We should be careful about that. Items can be equivalent (not lesser/greater) but not equal.
```suggestion
 * Equality is determined by comparing rows element by element. The first mismatching element
 * returns false, representing unequal rows. If the rows are compared without mismatched elements,
 * the rows are equal.
```